sdk/identity/identity/src/credentials/interactiveBrowserCredential.ts

```diff
@@ -116,71 +120,101 @@ export class InteractiveBrowserCredential implements TokenCredential {
-            logger.getToken.info(formatSuccess(scopeArray));
-
-            resolve({
-              expiresOnTimestamp,
-              token: authResponse.accessToken
-            });
-          } else {
+        this.msalClient
+          .acquireTokenByCode(tokenRequest)
+          .then((authResponse) => {
+            const successMessage = `Authentication Complete. You can close the browser and return to the application.`;
+            if (authResponse && authResponse.expiresOn) {
+              const expiresOnTimestamp = authResponse?.expiresOn.valueOf();
+              res.writeHead(200);
+              res.end(successMessage);
+              logger.getToken.info(formatSuccess(scopeArray));
+
+              resolve({
+                expiresOnTimestamp,
+                token: authResponse.accessToken
+              });
+            } else {
+              const errorMessage = formatError(
+                scopeArray,
+                `${url.searchParams.get("error")}. ${url.searchParams.get("error_description")}`
+              );
+              res.writeHead(500);
+              res.end(errorMessage);
+              logger.getToken.info(errorMessage);
+
+              reject(
+                new Error(
+                  `Interactive Browser Authentication Error "Did not receive token with a valid expiration"`
+                )
+              );
+            }
+            cleanup();
+            return;
+          })
+          .catch(() => {
+            const errorMessage = formatError(
+              scopeArray,
+              `${url.searchParams.get("error")}. ${url.searchParams.get("error_description")}`
+            );
+            res.writeHead(500);
+            res.end(errorMessage);
+            logger.getToken.info(errorMessage);
+
             reject(
               new Error(
                 `Interactive Browser Authentication Error "Did not receive token with a valid expiration"`
               )
             );
-          }
-        } catch (error) {
-          const errorMessage = formatError(
-            scopeArray,
-            `${req.query["error"]}. ${req.query["error_description"]}`
-          );
-          res.status(500).send(errorMessage);
-          logger.getToken.info(errorMessage);
-          reject(new Error(errorMessage));
-        } finally {
-          cleanup();
-        }
-      });
+            cleanup();
+          });
+      };
+      const app = http.createServer(requestListener);
 
-      listen = app.listen(this.port, () =>
+      const listen = app.listen(this.port, this.hostname, () =>
         logger.info(`InteractiveBrowerCredential listening on port ${this.port}!`)
       );
-      listen.on("connection", (socket) => (socketToDestroy = socket));
+      app.on("connection", (socket) => (socketToDestroy = socket));
+      const server = stoppable(app);
 
       try {
         await this.openAuthCodeUrl(scopeArray);
       } catch (e) {
         cleanup();
         throw e;
       }
```

Comment:
It's not super clear to me if this will do the right thing, since we're in a promise constructor asynchronously. I think it'd be safer to rewrite this bit and ditch the `async` from the Promise constructor callback:

```suggestion
      this.openAuthCodeUrl(scopeArray).catch(e => {
          cleanup();
          reject(e);
      });
```